tests/src/Unit/PermissionEventTest.php

```diff
@@ -0,0 +1,461 @@
+   *   An array of test permissions.
+   * @param string $entity_type_id
+   *   The entity type ID of the group type to which the permissions apply.
+   * @param string $bundle_id
+   *   The bundle ID of the group type to which the permissions apply.
+   *
+   * @covers ::hasPermission
+   *
+   * @dataProvider permissionsProvider
+   */
+  public function testHasPermission($permissions, $entity_type_id, $bundle_id) {
+    $event = new PermissionEvent($entity_type_id, $bundle_id);
+
+    foreach ($permissions as $name => $permission) {
+      $this->assertFalse($event->hasPermission($name));
+      $event->setPermission($name, $permission);
+      $this->assertTrue($event->hasPermission($name));
+    }
+  }
+
+  /**
+   * @param array $permissions
+   *   An array of test permissions.
+   * @param string $entity_type_id
+   *   The entity type ID of the group type to which the permissions apply.
+   * @param string $bundle_id
+   *   The bundle ID of the group type to which the permissions apply.
+   *
+   * @covers ::getEntityTypeId
+   *
+   * @dataProvider permissionsProvider
+   */
+  public function testGetEntityTypeId($permissions, $entity_type_id, $bundle_id) {
+    $event = new PermissionEvent($entity_type_id, $bundle_id);
+    $this->assertEquals($entity_type_id, $event->getEntityTypeId());
+  }
+
+  /**
+   * @param array $permissions
+   *   An array of test permissions.
+   * @param string $entity_type_id
+   *   The entity type ID of the group type to which the permissions apply.
+   * @param string $bundle_id
+   *   The bundle ID of the group type to which the permissions apply.
+   *
+   * @covers ::getBundleId
+   *
+   * @dataProvider permissionsProvider
+   */
+  public function testGetBundleId($permissions, $entity_type_id, $bundle_id) {
+    $event = new PermissionEvent($entity_type_id, $bundle_id);
+    $this->assertEquals($bundle_id, $event->getBundleId());
+  }
+
+  /**
+   * @param array $permissions
+   *   An array of test permissions.
+   * @param string $entity_type_id
+   *   The entity type ID of the group type to which the permissions apply.
+   * @param string $bundle_id
+   *   The bundle ID of the group type to which the permissions apply.
+   *
+   * @covers ::filterByDefaultRole
+   *
+   * @dataProvider permissionsProvider
+   */
+  public function testFilterByDefaultRole($permissions, $entity_type_id, $bundle_id) {
+    $event = new PermissionEvent($entity_type_id, $bundle_id);
+    $event->setPermissions($permissions);
+
+    $default_roles = [
+      OgRoleInterface::ANONYMOUS,
+      OgRoleInterface::AUTHENTICATED,
+      OgRoleInterface::ADMINISTRATOR,
+    ];
+    foreach ($default_roles as $default_role) {
+      $expected = array_filter($permissions, function ($permission) use ($default_role) {
+        return !empty($permission['default roles']) && in_array($default_role, $permission['default roles']);
+      });
+      $this->assertEquals($expected, $event->filterByDefaultRole($default_role));
```

Comment:
I know :( I am a bit torn between using them or going for KernelTests. For simple classes that are mainly intended to store data unit tests make perfect sense. For things like services that are depending on multiple injected dependencies it is a real chore to write them, and the end result is a test that runs fast but is very brittle.

I must say though that the unit tests give a new perspective on the code that was written and will point out inefficiencies.